modules/objdetect/src/aruco/aruco_detector.cpp

```diff
@@ -847,15 +665,232 @@ struct ArucoDetector::ArucoDetectorImpl {
                       const RefineParameters& _refineParams): dictionary(_dictionary),
                       detectorParams(_detectorParams), refineParams(_refineParams) {}
 
-    float getAverageArucoPinSize(vector<Point2f> markerCorners) {
+    float getAverageModuleSize(vector<Point2f> markerCorners) {
         float averageArucoModuleSize = 0.f;
         int numPins = dictionary.markerSize + detectorParams.markerBorderBits * 2;
         for (size_t i = 0ull; i < markerCorners.size(); i++) {
             averageArucoModuleSize += sqrt(normL2Sqr<float>(Point2f(markerCorners[i] - markerCorners[(i+1ull)%markerCorners.size()])));
         }
         averageArucoModuleSize /= ((float)markerCorners.size()*numPins);
         return averageArucoModuleSize;
-}
+    }
+
+    /**
+     * @brief Detect square candidates in the input image
+     */
+    void detectCandidates(const Mat& grey, vector<vector<Point2f> >& candidates, vector<vector<Point> >& contours) {
+        /// 1. DETECT FIRST SET OF CANDIDATES
+        _detectInitialCandidates(grey, candidates, contours, detectorParams);
+        /// 2. SORT CORNERS
+        _reorderCandidatesCorners(candidates);
+    }
+
+    /**
+     * @brief  FILTER OUT NEAR CANDIDATE PAIRS
+     * 
+     * save the outter/inner border (i.e. potential candidates) to vector<MarkerCandidateTree>,
+     * clear candidates and contours
+     */
+    vector<MarkerCandidateTree>
+    filterTooCloseCandidates(vector<vector<Point2f> > &candidates, vector<vector<Point> > &contours) {
+        CV_Assert(detectorParams.minMarkerDistanceRate >= 0);
+        vector<MarkerCandidateTree> candidateTree(candidates.size());
+        for(unsigned int i = 0; i < candidates.size(); i++) {
+            candidateTree[i] = MarkerCandidateTree(std::move(candidates[i]), std::move(contours[i]),
+                                                   dictionary.markerSize, detectorParams.markerBorderBits);
+        }
+        candidates.clear();
+        contours.clear();
+        std::sort(candidateTree.begin(), candidateTree.end());
+        // group index for each candidate
+        vector<int> groupId(candidateTree.size(), -1);
+        vector<vector<unsigned int> > groupedCandidates;
+        vector<bool> isSelectedContours(candidateTree.size(), true);
+        const float moduleGroupRate = (float)dictionary.markerSize*.5f + 1.f;
+                                      // (float)detectorParams.minMarkerDistanceRate*4.f*
+                                      // (dictionary.markerSize+2*detectorParams.markerBorderBits); // TODO: optimize parameter
+        const float moduleRate = .21f; // TODO: optimize parameter?
+
+        int countSelectedContours = 0;
+        for (unsigned int i = 0; i < candidateTree.size(); i++) {
+            for (unsigned int j = i + 1; j < candidateTree.size(); j++) {
```

Comment:
it makes sense to use size_t to prevent build warnings.

Reply:
added size_t